### PR TITLE
feat: mayachain swapper support

### PR DESCRIPTION
**Pull request by kaladinlight**
## Description

Support swapping with Mayachain.

Note:
- It is currently difficult to parse thor vs maya swaps without a maya specific affiliate name. Follow up with appropriate logic once affiliate name is created.
- ~~I need to track down a precision issue with CACAO swaps as a last missing piece, but other swaps are working as intended.~~

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/9235

## Risk

> High Risk PRs Require 2 approvals

Medium - generally speaking this is a new feature, but there is some associated refactoring that deserves a solid regression test.

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

- Ensure you can swap via mayachain with supported assets (known cacao precision issue at the moment)
- High level regression test of thorchain swaps

### Engineering

:point_up:

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

:point_up:

## Screenshots (if applicable)

image

image

### CODE REVIEW

File: packages/swapper/src/thorchain-utils/memo/assertAndProcessMemo.ts [hunk 1]
```diff
@@ -141,7 +141,9 @@ export const assertAndProcessMemo = (memo: string, affiliate: string): string =>
         return ''
       })()
 
-      return `${_action}:${asset}:${destAddr}:${limit}:${affiliate}:${fee || 0}${maybeSwapOutParts}`
+      return `${_action}:${asset}:${destAddr}:${limit}:${affiliate}:${
+        affiliate ? fee || 0 : ''
```
Comment (kaladinlight) [inline packages/swapper/src/thorchain-utils/memo/assertAndProcessMemo.ts:145]: currently handles no affiliate name, can revert after it is created and in use

File: packages/unchained-client/src/evm/ethereum/parser/__tests__/ethereum.test.ts [hunk 2]
```diff
@@ -1901,6 +1903,47 @@ describe('parseTx', () => {
     })
   })
 
+  describe('mayachain', () => {
+    it.skip('should be able to parse token swap', async () => {
```
Comment (kaladinlight) [inline packages/unchained-client/src/evm/ethereum/parser/__tests__/ethereum.test.ts:1907]: todo after affiliate name and futher parsing edge cases are resolved

File: packages/unchained-client/src/parser/thorchain.ts [hunk 3]
```diff
@@ -119,69 +122,77 @@ const getLiquidityType = (pool: string): LiquidityType => (pool.includes('/') ?
 const getSwapType = (memo: string): SwapType =>
   streamingSwapRegex.test(memo) ? 'Streaming' : 'Standard'
 
-export class Parser {
+export class Parser<T extends 'thorchain' | 'mayachain' = 'thorchain'> {
```
Comment (kaladinlight) [inline packages/unchained-client/src/parser/thorchain.ts:125]: could really use a base class, but I didn't have a good name for it so just went with this for now

File: src/react-queries/queries/thornode.ts [hunk 4]
```diff
@@ -1,81 +1,106 @@
...
-    }
-  },
-})
+  })()
+
+  return {
+    queryKey,
+    queryFn: queryFn as QueryFunction<ThorchainMimir>,
+  }
+}
```
Comment (kaladinlight) [inline src/react-queries/queries/thornode.ts:106]: This was the most ergonomic I could manage to allow for easiest possible conditional calls against thorchain or mayachain given a `chainId`. Open to feedback though.

File: src/react-queries/queries/thornode.ts [hunk 5]
```diff
@@ -1,81 +1,106 @@
...
-      return data
-    },
-  }),
-  poolsData: () => ({
-    queryKey: ['thornodePoolsData'],
-    queryFn: async () => {
-      const poolResponse = await thorService.get<ThornodePoolResponse[]>(
-        `${thornodeUrl}/thorchain/pools`,
-      )
+export const getInboundAddressesQuery = (chainId: ChainId) => {
```
Comment (kaladinlight) [inline src/react-queries/queries/thornode.ts:72]: This was the most ergonomic I could manage to allow for easiest possible conditional calls against thorchain or mayachain given a chainId. Open to feedback though.

**Review by 0xApotheosis — COMMENTED** [review]
I was just using this branch to get some CACAO and noticed that the recipient address on the "Sign and Swap" screen tries to use Etherscan to show the Mayachain account:

Screenshot 2025-06-04 at 4 17 26 pm

https://etherscan.io/address/maya1j5jxvwd33rz66r88vwwwayvncjadx9jy5qjqk7

File: headers/csps/chains/mayachain.ts [hunk 6]
```diff
@@ -9,7 +9,7 @@ export const csp: Csp = {
   'connect-src': [
     env.VITE_UNCHAINED_MAYACHAIN_HTTP_URL,
     env.VITE_UNCHAINED_MAYACHAIN_WS_URL,
-    env.VITE_MAYACHAIN_NODE_URL,
+    `${env.VITE_MAYACHAIN_NODE_URL}/`,
```
Comment (gomesalexandre) [inline headers/csps/chains/mayachain.ts:12]: eh, why not!

File: packages/chain-adapters/src/evm/arbitrum/ArbitrumChainAdapter.ts [hunk 7]
```diff
@@ -29,6 +33,7 @@ export class ChainAdapter extends EvmBaseAdapter<KnownChainIds.ArbitrumMainnet>
         chainId: args.chainId ?? DEFAULT_CHAIN_ID,
         rpcUrl: args.rpcUrl,
         api: args.providers.http,
+        mayaMidgardUrl: args.mayaMidgardUrl,
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/evm/arbitrum/ArbitrumChainAdapter.ts:36]: Any way we could do this without explicitly having `mayaMidgardUrl` in every single adapter?

File: packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts [hunk 8]
```diff
@@ -131,7 +139,7 @@ export const getL1RateOrQuote = async <T extends ThorTradeRateOrQuote>(
   const recommendedMinimumCryptoBaseUnit = swapQuote.recommended_min_amount_in
     ? convertPrecision({
         value: swapQuote.recommended_min_amount_in,
-        inputExponent: nativePrecision,
+        inputExponent: sellAssetNativePrecision,
```
Comment (gomesalexandre) [inline packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts:145]: Naming doesn't look correct - this seems to imply that this is the precision of the native asset for the sell asset - in reality, it's the native precision of the either the *fee* asset for the chain (i.e RUNE/CACAO precision), *or* MAYA right?

File: packages/swapper/src/thorchain-utils/index.ts [hunk 9]
```diff
@@ -74,12 +87,12 @@ export const isNativeAsset = (assetId: AssetId, swapperName: SwapperName) => {
   }
 }
 
-export const getNativePrecision = (swapperName: SwapperName) => {
+export const getNativePrecision = (assetId: AssetId, swapperName: SwapperName) => {
   switch (swapperName) {
     case SwapperName.Thorchain:
-      return THOR_PRECISION
+      return THORCHAIN_PRECISION
```
Comment (gomesalexandre) [inline packages/swapper/src/thorchain-utils/index.ts:93]: q: should this handle TCY similar to the below?

File: packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts (hunk 8 above)
Comment (gomesalexandre) [reply packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts:145]: Although I'm not sure I can come with something better here so meh!

File: packages/swapper/src/thorchain-utils/memo/assertAndProcessMemo.ts [hunk 10]
```diff
@@ -62,18 +62,18 @@ function assertMemoHasAction(action: string | undefined, memo: string): asserts
 const assertIsValidLimit = (limit: string | undefined, memo: string) => {
   assertMemoHasLimit(limit, memo)
 
-  const maybeStreamingSwap = limit.match(/\//g)
-  const [lim, interval, quantity] = limit.split('/')
+  const parts = limit.split('/')
```
Comment (gomesalexandre) [inline packages/swapper/src/thorchain-utils/memo/assertAndProcessMemo.ts:65]: q: shouldn't this be `streamingParts` ?

File: packages/utils/src/assetData/baseAssets.ts [hunk 11]
```diff
@@ -147,7 +147,7 @@ export const maya: Readonly<Asset> = Object.freeze({
   symbol: 'MAYA',
   precision: 4,
   color: '#00FFA0',
-  icon: 'https://raw.githubusercontent.com/shapeshift/web/feat_maya_asset/scripts/generateAssetData/thorchain/icons/maya.png',
+  icon: 'https://raw.githubusercontent.com/shapeshift/web/develop/scripts/generateAssetData/thorchain/icons/maya.png',
```
Comment (gomesalexandre) [inline packages/utils/src/assetData/baseAssets.ts:150]: 💜

File: src/lib/getTxLink.ts [hunk 12]
```diff
@@ -53,12 +55,15 @@ export const getTxBaseUrl = ({
     case THORCHAIN_STREAM_SWAP_SOURCE:
     case THORCHAIN_LONGTAIL_SWAP_SOURCE:
     case THORCHAIN_LONGTAIL_STREAMING_SWAP_SOURCE:
-      return 'https://viewblock.io/thorchain/tx/'
+      return thorchain.explorerTxLink
```
Comment (gomesalexandre) [inline src/lib/getTxLink.ts:58]: 💜

File: src/lib/utils/thorchain/hooks/useThorchainMimir.tsx
Comment (gomesalexandre) [inline src/lib/utils/thorchain/hooks/useThorchainMimir.tsx:1]: Looks like this could be kept as .tsx?

File: src/plugins/litecoin/index.tsx [hunk 13]
```diff
@@ -30,7 +30,7 @@ export default function register(): Plugins {
                 return new litecoin.ChainAdapter({
                   providers: { http, ws },
                   coinName: 'Litecoin',
-                  midgardUrl: getConfig().VITE_THORCHAIN_MIDGARD_URL,
+                  thorMidgardUrl: getConfig().VITE_THORCHAIN_MIDGARD_URL,
```
Comment (gomesalexandre) [inline src/plugins/litecoin/index.tsx:33]: same q here, curious why some plugins have both and some e.g this have only the THOR one?

File: src/react-queries/queries/thornode.ts (hunk 5 above)
Comment (gomesalexandre) [reply src/react-queries/queries/thornode.ts:72]: Honestly we should probably ditch those queries altogether and keep them regular functions, which would make things a lot easier.
Rationale is in effect, we more often than not consume the queryKey and queryFn from these separately, as opposed to consuming those abstracted queries as intended, so we can leverage the `criteria ? queryFn : skipToken` logic.

Not a now thing anyway as that would be quite a change on top of this monster of a PR, so totally happy with this for the time being but just something to keep in mind if we feel like we want to clean this up later at some point! 🙏🏽

Comment (gomesalexandre) [reply src/react-queries/queries/thornode.ts:72]: ^ note the comment above refers to `createQueryKeys()` bits - as far as this and `getMimirQuery` are concerned, this is perfect IMO!

File: src/state/helpers.ts [hunk 14]
```diff
@@ -65,7 +65,7 @@ export const getEnabledSwappers = (
     [SwapperName.Relay]:
       RelaySwapper && (!isCrossAccountTrade || isCrossAccountTradeSupported(SwapperName.Relay)),
     [SwapperName.Mayachain]:
-      Mayachain && (!isCrossAccountTrade || isCrossAccountTradeSupported(SwapperName.Mayachain)),
+      MayaSwap && (!isCrossAccountTrade || isCrossAccountTradeSupported(SwapperName.Mayachain)),
```
Comment (gomesalexandre) [inline src/state/helpers.ts:68]: 💜

File: src/state/slices/common/tradeInputBase/createTradeInputBaseSlice.ts [hunk 15]
```diff
@@ -25,7 +24,6 @@ const getBaseReducers = <T extends TradeInputBaseState>(initialState: T) => ({
   clear: () => initialState,
   setBuyAsset: (state: Draft<T>, action: PayloadAction<Asset>) => {
     const asset = action.payload
-    if (asset.chainId === KnownChainIds.MayachainMainnet) return
```
Comment (gomesalexandre) [inline src/state/slices/common/tradeInputBase/createTradeInputBaseSlice.ts:28]: 👋🏽

**Review by gomesalexandre — COMMENTED** [review]
A few comments but looks good! The split between THOR and MAYA looks a lot cleaner in this PR 🐐 

Runtime test incoming 🙏🏽

File: headers/csps/chains/mayachain.ts (hunk 6 above)
Comment (kaladinlight) [reply headers/csps/chains/mayachain.ts:12]: just copied from thorchain csp. wasn't working without it with the addition of `/lcd` to the path :shrug:

File: packages/chain-adapters/src/evm/arbitrum/ArbitrumChainAdapter.ts (hunk 7 above)
Comment (kaladinlight) [reply packages/chain-adapters/src/evm/arbitrum/ArbitrumChainAdapter.ts:36]: I mean I suppose we can just fetch from process.env directly in unchained-client, but this was just the pattern of passing in args to initialize the transaction parser. They are only added to chains where thorchain/mayachain are supported.

File: packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts (hunk 8 above)
Comment (kaladinlight) [reply packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts:145]: So it is intended on representing the precision of the asset in term of Thorchain/Mayachain base units. This amount to precision of 8 for everything except for CACAO which is precision 10. So I have been using "native" to imply the network base units if you will. I didn't want it to mean the actual precision on the asset as it may exist on it's own network however which certainly could be confused with the "native" notion.

I can add a comment at minimum for explanation, but there really isn't a great name outside of thormayaPrecision or something stupid like that lol

File: packages/swapper/src/thorchain-utils/index.ts (hunk 9 above)
Comment (kaladinlight) [reply packages/swapper/src/thorchain-utils/index.ts:93]: AFAIK TCY shares the same 8 decimal precision as the rest of the thorchain assets. CACAO is the only weird one at 10 decimals.

File: packages/swapper/src/thorchain-utils/memo/assertAndProcessMemo.ts (hunk 10 above)
Comment (kaladinlight) [reply packages/swapper/src/thorchain-utils/memo/assertAndProcessMemo.ts:65]: it would be limit or streaming parts technically. parts.length === 1 implies just limit (standard swap), parts.length === 3 would be streaming swap. I just went with general naming here, we are splitting on `/` and getting back some number of "parts"

File: src/lib/utils/thorchain/hooks/useThorchainMimir.tsx
Comment (kaladinlight) [reply src/lib/utils/thorchain/hooks/useThorchainMimir.tsx:1]: the `tsx` was doing weird things with the linter with the generics that was irritating me lol
image

File: src/plugins/litecoin/index.tsx (hunk 13 above)
Comment (kaladinlight) [reply src/plugins/litecoin/index.tsx:33]: It relates to thorchain/mayachain support on that chain. Right now all of the utxo chains are supported on thorchain, but not all utxo chains are supported on mayachain.

**Comment by kaladinlight** [comment]
> Tested locally, LGTM! @kaladinlight one q. related to swaps being displayed as streaming in MAYA explorer, though non-blocking and perhaps not an us issue, happy to investigate in a follow-up 🙏🏽
> 
> ### THOR
> * Long-tail -> L1 are borked but already borked in prod https://jam.dev/c/d1f190f9-c8ef-44c2-8d1f-a73b0b70d253
> * L1 -> L1 is happy https://jam.dev/c/1654de34-b3ca-4156-a98b-9ec69ef4060d
> * L1 -> RUNE is happy https://jam.dev/c/67e27ccb-3f46-4de8-8879-092359e814f1
> 
> ### MAYA
> * RUNE -> CACAO is happy ❓  https://www.mayascan.org/tx/E74B7121DEEF84F33DFB0FBD6B5E9851C0F872782FF9F5DC826F2738A447C2C6 @kaladinlight not sure whether that's an us thing and whether that's blocking, but Maya explorer is displaying this as a 1 interval streaming swap, is this expected?
> * L1 -> CACAO is happy https://jam.dev/c/31fb3c26-8167-4742-ae41-ba98bebc0e98
> * CACAO -> L1 is happy https://jam.dev/c/d9e644b7-f231-4a5e-adeb-62970ebb6f45

This looks to be the case across the board on mayascan. Not an us thing though and our memos will look cleaner without the trailing `::` after affiliate name is integrated

File: packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts (hunk 8 above)
Comment (gomesalexandre) [reply packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts:145]: Happy with a comment @kaladinlight !

Comment (kaladinlight) [reply packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts:145]: 00e5c6fb576c5758cfcb5c99a82108a98156efe1
